### PR TITLE
feat(email): add support for pgp encrypting notifications

### PR DESCRIPTION
**Pull request by ankarhem**
#### Description
Add a field to the user profile for adding PGP Key. If it exists it will be used to encrypt the email notification sent when sending email notifications to the user.

#### To-Dos
- [x] Successful build `yarn build`
- [x] Translation keys `yarn i18n:extract`
- [x] Database migration (if required)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -59,6 +60,7 @@
     "reflect-metadata": "^0.1.13",
     "secure-random-password": "^0.2.2",
     "sqlite3": "^5.0.2",
+    "stream": "^0.0.2",
```
Comment (sct) [inline package.json:63]: I don't think we need to add `stream` as a package. It's built into node?

File: src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx [hunk 2]
```diff
@@ -32,6 +32,9 @@ const messages = defineMessages({
   localuser: 'Local User',
   toastSettingsSuccess: 'Settings successfully saved!',
   toastSettingsFailure: 'Something went wrong while saving settings.',
+  pgpKey: 'PGP Public Key',
+  pgpKeyTip:
+    'By setting this key, any emails sent to you will be encrypted with this key.',
```
Comment (TheCatLady) [inline src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx:37]:
```suggestion
    'Encrypt email messages using <PgpLink>PGP</PgpLink>',
```

We could link to https://en.wikipedia.org/wiki/Pretty_Good_Privacy perhaps? 😃

Also, do we need to do any input validation on the PGP key to ensure that users who don't understand what it is don't enter random things and complain that their email notifications stopped working?

File: src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx [hunk 3]
```diff
@@ -123,6 +128,22 @@ const UserNotificationSettings: React.FC = () => {
                   />
                 </div>
               </div>
+              <div className="form-row">
+                <label htmlFor="pgpKey" className="text-label">
+                  <span>{intl.formatMessage(messages.pgpKey)}</span>
```
Comment (TheCatLady) [inline src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx:133]:
```suggestion
                  <span>{intl.formatMessage(messages.pgpKey)}</span>
                  <Badge badgeType="danger">
                      {intl.formatMessage(globalMessages.advanced)}
                    </Badge>
```

I think we should mark this field as "Advanced," similar to the "Enable CSRF Protection" setting.

File: src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx [hunk 4]
```diff
@@ -32,6 +34,8 @@ const messages = defineMessages({
   localuser: 'Local User',
   toastSettingsSuccess: 'Settings successfully saved!',
   toastSettingsFailure: 'Something went wrong while saving settings.',
+  pgpKey: 'PGP Public Key',
+  pgpKeyTip: 'Encrypt email messages using <PgpLink>PGP</PgpLink>',
```
Comment (TheCatLady) [inline src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx:38]:
```suggestion
  pgpKey: '<PgpLink>PGP</PgpLink> Public Key',
  pgpKeyTip: 'Encrypt email messages',
```

File: src/components/Settings/Notifications/NotificationsEmail.tsx [hunk 5]
```diff
@@ -36,6 +38,12 @@ const messages = defineMessages({
   emailNotificationTypesAlertDescription:
     'For the "Media Requested" and "Media Failed" notification types,\
     notifications will only be sent to users with the "Manage Requests" permission.',
+  pgpPrivateKey: 'PGP Private Key',
+  pgpPrivateKeyTip:
+    'Sign encrypted email messages (PGP password is also required)',
+  pgpPassword: 'PGP Password',
+  pgpPasswordTip:
+    'Sign encrypted email messages (PGP private key is also required)',
```
Comment (TheCatLady) [inline src/components/Settings/Notifications/NotificationsEmail.tsx:46]:
```suggestion
  pgpPrivateKey: '<PgpLink>PGP</PgpLink> Private Key',
  pgpPrivateKeyTip:
    'Sign encrypted email messages (PGP password is also required)',
  pgpPassword: '<PgpLink>PGP</PgpLink> Password',
  pgpPasswordTip:
    'Sign encrypted email messages (PGP private key is also required)',
```

File: src/components/Settings/Notifications/NotificationsEmail.tsx [hunk 6]
```diff
@@ -265,6 +279,52 @@ const NotificationsEmail: React.FC = () => {
                   </div>
                 </div>
               </div>
+              <div className="form-row">
+                <label htmlFor="pgpPrivateKey" className="text-label">
+                  <span className="mr-2">
+                    {intl.formatMessage(messages.pgpPrivateKey)}
```
Comment (TheCatLady) [inline src/components/Settings/Notifications/NotificationsEmail.tsx:285]:
```suggestion
                    {intl.formatMessage(messages.pgpPrivateKey, {
                      PgpLink: function PgpLink(msg) {
                        return (
                          <a
                            href="https://en.wikipedia.org/wiki/Pretty_Good_Privacy"
                            target="_blank"
                            rel="noreferrer"
                            className="text-gray-100 underline transition duration-300 hover:text-white"
                          >
                            {msg}
                          </a>
                        );
                      },
                    })}
```

File: src/components/Settings/Notifications/NotificationsEmail.tsx [hunk 7]
```diff
@@ -265,6 +279,52 @@ const NotificationsEmail: React.FC = () => {
...
+                      as="textarea"
+                      rows="3"
+                    />
+                  </div>
+                </div>
+              </div>
+              <div className="form-row">
+                <label htmlFor="pgpPassword" className="text-label">
+                  <span className="mr-2">
+                    {intl.formatMessage(messages.pgpPassword)}
```
Comment (TheCatLady) [inline src/components/Settings/Notifications/NotificationsEmail.tsx:308]:
```suggestion
                    {intl.formatMessage(messages.pgpPassword, {
                      PgpLink: function PgpLink(msg) {
                        return (
                          <a
                            href="https://en.wikipedia.org/wiki/Pretty_Good_Privacy"
                            target="_blank"
                            rel="noreferrer"
                            className="text-gray-100 underline transition duration-300 hover:text-white"
                          >
                            {msg}
                          </a>
                        );
                      },
                    })}
```

File: src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx [hunk 8]
```diff
@@ -123,6 +129,40 @@ const UserNotificationSettings: React.FC = () => {
...
+                            href="https://en.wikipedia.org/wiki/Pretty_Good_Privacy"
+                            target="_blank"
+                            rel="noreferrer"
+                            className="text-gray-100 underline transition duration-300 hover:text-white"
+                          >
+                            {msg}
+                          </a>
+                        );
+                      },
+                    })}
```
Comment (TheCatLady) [inline src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx:154]:
```suggestion
                    {intl.formatMessage(messages.pgpKeyTip)}
```

File: src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx [hunk 9]
```diff
@@ -123,6 +129,40 @@ const UserNotificationSettings: React.FC = () => {
                   />
                 </div>
               </div>
+              <div className="form-row">
+                <label htmlFor="pgpKey" className="text-label">
+                  <span className="mr-2">
+                    {intl.formatMessage(messages.pgpKey)}
```
Comment (TheCatLady) [inline src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx:135]:
```suggestion
                    {intl.formatMessage(messages.pgpKey, {
                      PgpLink: function PgpLink(msg) {
                        return (
                          <a
                            href="https://www.openpgp.org/"
                            target="_blank"
                            rel="noreferrer"
                            className="text-gray-100 underline transition duration-300 hover:text-white"
                          >
                            {msg}
                          </a>
                        );
                      },
                    })}
```

**Review by JHilmarch — COMMENTED** [review]
Fancy stuff :) Snowden is proud of you!

I was concentrating on the readability of the code. Please read my comments and consider my suggestions for improvements.

When going to beta you can consider including the PgpEncrypter into a scenario-test:
1. Generate a real world email
2. Encrypt it
3. Decrypt it
4. Assert generated and decrypted email content is the same

File: src/components/Settings/Notifications/NotificationsEmail.tsx [hunk 10]
```diff
@@ -265,6 +279,52 @@ const NotificationsEmail: React.FC = () => {
...
+                  </span>
+                  <Badge badgeType="danger">
+                    {intl.formatMessage(globalMessages.advanced)}
+                  </Badge>
+                  <span className="label-tip">
+                    {intl.formatMessage(messages.pgpPrivateKeyTip)}
+                  </span>
+                </label>
+                <div className="form-input">
+                  <div className="flex max-w-lg rounded-md shadow-sm">
```
Comment (TheCatLady) [inline src/components/Settings/Notifications/NotificationsEmail.tsx:318]:
```suggestion
                  <div className="form-input-field">
```

File: src/components/Settings/Notifications/NotificationsEmail.tsx [hunk 11]
```diff
@@ -265,6 +279,52 @@ const NotificationsEmail: React.FC = () => {
...
```
Comment (TheCatLady) [inline src/components/Settings/Notifications/NotificationsEmail.tsx:295]:
```suggestion
                  <div className="form-input-field">
```

File: src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx [hunk 12]
```diff
@@ -123,6 +129,40 @@ const UserNotificationSettings: React.FC = () => {
...
+                          >
+                            {msg}
+                          </a>
+                        );
+                      },
+                    })}
+                  </span>
+                </label>
+                <div className="form-input">
+                  <div className="flex max-w-lg rounded-md shadow-sm">
```
Comment (TheCatLady) [inline src/components/UserProfile/UserSettings/UserNotificationSettings/index.tsx:158]:
```suggestion
                  <div className="form-input-field">
```

**Review by TheCatLady — COMMENTED** [review]
Use new `form-input-field` global CSS class (added in 1be97fe7fb4ef6bf86c9874cb90c5deb94b2509a)

**Review by sct — APPROVED** [review]

**Pull request merged**